### ISSUE TITLE
Printer instance has no attribute 'try_load_module'

### ISSUE THREAD

**Issue by markwinger** [Stephan3/dwc2-for-klipper#73]
I have been trying to install dwc2 for 2 days.  I was running octoprint with klipper and tried installing as described at:

https://github.com/Stephan3/dwc2-for-klipper

but that fails at wget https://github.com/chrishamm/DuetWebControl/releases/download/2.1.6/DuetWebControl-SBC.zip
 with the error indicating 2.1.6 does not exist.  I tried using:
wget https://github.com/chrishamm/DuetWebControl/releases/download/2.0.0-RC5/DuetWebControl.zip

which is described at:https://klipper.info/klipper-+-dwc2-1/installing-dwc2

That installed but the gui would not come up.  If I open octoprint the klipper  console "status" says:

Internal error during connect: Printer instance has no attribute 'try_load_module'

I even tried installing on a clean raspbian system with the installer.sh  descibed at:https://github.com/manu7irl/klipper-DWC2-installer?files=1.
I only have one printer so this is more that I need but I tried it anyway.  
With this I could not get dwc2, or octoprint guis to open and of course klipper then is not running.

I did install this a few months ago and it worked but I decided not to switch then so went back to octoprint. (no longer installed).

**The dwc2 gui fails to open(Unable to connect), and klipper does not start with this try_load_module error.**  

What next????

**Comment by manu7irl** [comment]
The problem is related to the big klipper commit that has been done a few
days ago.
It made some changes to the gcode.py file with that the change that Stephan
has made in that file to make dwc2 working are not inline with the new file
structure.
We need to revert to an older commit and wait for Stephan to make the
changes needed in order to get that fixed.

**Comment by markwinger** [comment]
On github 2.1.7 it showed a duetwebcontrol-SD.zip so I tried that too.  No difference.

**Comment by Stephan3** [comment]
@manu7irl you pointed the Klipper commit allready. I have not the time doing much theese days.

**Comment by manu7irl** [comment]
I will try to make the changes needed to your sed commands. I will try to
send a pr.

**Comment by markwinger** [comment]
Octoprint works with klipper if I comment out the [web_dwc2] section.

**Comment by BlackStump** [comment]
If you are desperate to try something other then Octoprint you could try 
[Mainsail](https://github.com/meteyou/mainsail)
It is alpha but seems to be quite functional, it has a DWC2 look about it.

**Comment by pajak15** [comment]
I had the same issue today while doing a fresh install. Try changing at line 49 in the web_dwc2.py "try_load_module" with "load_object". As stated before there was a commit in klipper where they renamed try_load_module() to load_object().

**Comment by pluuuk** [comment]
a simple sed command to replace try_load_module with load_object works.

**Comment by Szeker** [comment]
@pluuuk: Are you on the latest Klipper? Simply replacing the "try_load..." to "load_object" seems not enough. "gcode.py" also has been modified in the latest version, so the original patch probably needs to be updated also.

**Comment by pluuuk** [comment]
@Szeker , I opened #75 with the fixes to get dwc2 working again.

**Comment by markwinger** [comment]
I am away now but i will try the fix tomorrow and update.

Get BlueMail for Android<http://www.bluemail.me/r?b=15860>
On May 10, 2020, at 4:57 PM, pluuuk <notifications@github.com<mailto:notifications@github.com>> wrote:

@Szeker<https://github.com/Szeker> , I opened #74<https://github.com/Stephan3/dwc2-for-klipper/pull/74> with the fixes to get dwc2 working again.

**Comment by markwinger** [comment]
OK, one install says to use wget https://github.com/chrishamm/DuetWebControl/releases/download/2.1.6/DuetWebControl-SBC.zip.  But this does not exist, I get 2020-05-11 06:55:38 ERROR 404: Not Found.

The other: wget https://github.com/chrishamm/DuetWebControl/releases/download/2.0.0-RC5/DuetWebControl.zip is well over a year old.  

There are no SBC files after 2.1.4.  What do you recommend?

**Comment by manu7irl** [comment]
```wget https://github.com/chrishamm/DuetWebControl/releases/download/2.1.7/DuetWebControl-SD.zip```

**Comment by markwinger** [comment]
OK, that was the last one I pulled.  So just to clarify (I don't understand git) if I just repeat the install process these fixes will get installed?

**Comment by markwinger** [comment]
Ok, I tried again and installed using the method here: https://github.com/Stephan3/dwc2-for-klipper

Same problem.  I modified  klippy/extras/web_dwc2.py changing try_load_object to load_object.  But the klipper logs show:
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 49, in __init__
    self.stepper_enable = self.printer.try_load_module(config, "stepper_enable") 
This line is now at line 50.  I assume I need to compile it but I don't know how.  What should I do?

**Comment by manu7irl** [comment]
you need to clone @pluuuk folder instaed of Stephan because this work is not merged yet...
git clone https://github.com/pluuuk/dwc2-for-klipper

**Comment by markwinger** [comment]
Ok, still no joy.

I used the instructions on the page you pointed to, substituting pluuuk for Stephan3 on the clode of dwc2-for-klipper.git.  The install seems to work.  Octoprint no longer complains about try_load_module, but when I open ip:4750, no gui. (Unable to connect)

**Comment by manu7irl** [comment]
Use my script and report there I got it patched, AFAIK, it should work now.
https://github.com/manu7irl/klipper-dwc2-installer

**Comment by markwinger** [comment]
I just tried to install dwc2 with it and it complains that the PrinterFarm direcoty does not exist.  Do I need to unistall everything and install everything with you script to move it all to the PrinterFarm?

**Comment by pluuuk** [comment]
@markwinger, did you configure the virtual_sdcard and patched gcode.py using the sed commands?

**Comment by pluuuk** [comment]
@manu7irl, the dwc2-for-klipper repo is 7 months old. I'm not sure if @markwinger will be able to use your installer.

**Comment by pluuuk** [comment]
These are the ammended instructions to run the fork that fixes these issues:
```
sudo systemctl stop klipper
cd ~
mv klipper klipper_backup 
PYTHONDIR="${HOME}/klippy-env"
virtualenv ${PYTHONDIR}
${PYTHONDIR}/bin/pip install tornado==5.1.1

git clone https://github.com/KevinOConnor/klipper.git
git clone https://github.com/pluuuk/dwc2-for-klipper.git
ln -s ~/dwc2-for-klipper/web_dwc2.py ~/klipper/klippy/extras/web_dwc2.py

mkdir -p ~/sdcard/dwc2/web
mkdir -p ~/sdcard/sys
cd ~/sdcard/dwc2/web 
wget https://github.com/chrishamm/DuetWebControl/releases/download/2.1.6/DuetWebControl-SBC.zip
unzip *.zip && for f_ in $(find . | grep '.gz');do gunzip ${f_};done
sudo systemctl start klipper
```

You will still need to configure the virtual_sdcard as in the readme

**Comment by markwinger** [comment]
Yes:
[virtual_sdcard]
path: ~/.octoprint/uploads/

and yes
gcode=$(sed 's/self.bytes_read = 0/self.bytes_read = 0\n        self.respond_callbacks = []/g' klipper/klippy/gcode.py)
gcode=$(echo "$gcode" | sed 's/# Response handling/def register_respond_callback(self, callback):\n        self.respond_callbacks.append(callback)/')
gcode=$(echo "$gcode" | sed 's/os.write(self.fd, msg+"\\n")/os.write(self.fd, msg+"\\n")\n            for callback in self.respond_callbacks:\n                callback(msg+"\\n")/')
echo "$gcode" > klipper/klippy/gcode.py

What about PrinterFarm directory?  Do I have to uninstall klipper and resinstall it with your script?  Klipper is installed and running with octoprint and I would like to keep it and add dwc2 so I can use either as I did when I tried it a few months ago.  Can that be done with you installer.sh script?

**Comment by pluuuk** [comment]
@markwinger , please see my last comment, I changed some things in the fork to remove the need to run the sed commands to modify gcode.py. This should allow you to install this much easier. 
I'm not familiar with @manu7irl install scripts.

**Comment by markwinger** [comment]
This still fails:
wget https://github.com/chrishamm/DuetWebControl/releases/download/2.1.6/DuetWebControl-SBC.zip

I changed it to 2.1.7/DuetWebControl-SD.zip.

Installed ok, but klipper fails to start again with a new error as shown in octoprints klipper console:
Internal error during connect: GCodeParser instance has no attribute 'register_respond_callback'

**Comment by pluuuk** [comment]
make sure you're running a clean klipper install, it seems you're running a
modified gcode.py
```
cd klipper
git checkout -- klippy/gcode.py
```

Should help you clean it up.

**Comment by markwinger** [comment]
No change.  So I deleted gcode.py, then did checkout again.  Still no change.

**Comment by pluuuk** [comment]
Can you run this and see if you still see the issue?
```
  cd dwc2-for-klipper
  git checkout -- .
  git pull
````

**Comment by markwinger** [comment]
Now klipper starts but the dwc2 gui will not connect.

**Comment by pluuuk** [comment]
Can you attach to a gist a clean /tmp/klippy.log?  Make sure to issue a RESTART or FIRMWARE_RESTART command first.

**Comment by markwinger** [comment]
Rebooted.  Then attacked with octoprint, then tried to open dwc2 gui.
[klippy.log](https://github.com/Stephan3/dwc2-for-klipper/files/4612906/klippy.log)

**Comment by pluuuk** [comment]
It seems your SD card path is configured for octoprint and its not pointing to where you extracted duetwebcontrol.

This is the error:
`DWC2 failed to start, no webif found in /home/pi/.octoprint/uploads/dwc2/web`

**Comment by markwinger** [comment]
Thanks.  I had the path commented out and was using the one I had for octoprint.  The gui starts now.

**Comment by pluuuk** [comment]
Thanks, let me know if you find issues. The fixes haven't been fully tested
yet, I hope some more users get a chance to ensure it all works properly.

**Comment by manu7irl** [comment]
> Yes:
> [virtual_sdcard]
> path: ~/.octoprint/uploads/
> 
> and yes
> gcode=$(sed 's/self.bytes_read = 0/self.bytes_read = 0\n self.respond_callbacks = []/g' klipper/klippy/gcode.py)
> gcode=$(echo "$gcode" | sed 's/# Response handling/def register_respond_callback(self, callback):\n self.respond_callbacks.append(callback)/')
> gcode=$(echo "$gcode" | sed 's/os.write(self.fd, msg+"\n")/os.write(self.fd, msg+"\n")\n for callback in self.respond_callbacks:\n callback(msg+"\n")/')
> echo "$gcode" > klipper/klippy/gcode.py
> 
> What about PrinterFarm directory? Do I have to uninstall klipper and resinstall it with your script? Klipper is installed and running with octoprint and I would like to keep it and add dwc2 so I can use either as I did when I tried it a few months ago. Can that be done with you installer.sh script?

Hi my script is for scratch install. Save your printer.cfg file. Uninstall the regular klipper and reinstall it with my script klipper and dwc2 will then be installed in a structured folder called PrinterFarm. And everything will be handle from there, in Octoprint nothing is needed to change but the virtual serial endpoint which will be /tmp/printer-0 and the printer.cfg file location will be at ~/PrinterFarm/KlipperFarm/printer-0/printer-0.cfg.
If you uninstall command to remove the old init.d startup script tell me I will give you.
For follow my readme to understand more about the changes.

**Comment by markwinger** [comment]
I did an overnight printer with it.  The printer came out fine but there were many notifications on the screen and I can't get rid of them.  They say:
ok B27.1/0.0 TO 25.0/0.4
Not SD Printing
ok

or
SD Not Printing
ok
ok B26:26.0/0.0 to B26:0/0.0

These messages messages all show up on the klipper console in too with a lot of NOT SD printing ok messages.  These messages seem to never stop.  The print finished 5 or six hours ago but these popup messages seem to still be printing.   There are so many the cover the right side of the screen top to bottom, covering other important info.

How do I stop them?  Can I turn off these notification popups somehow, or filter all these constant status messages.

**Comment by manu7irl** [comment]
I am almost 10 hours printing without issue and no sdcard messages.

**Comment by markwinger** [comment]
I also cannot run any macros from the dashboard/macros or the file management/macros.
If I specify a macro in the file management/macros I get an error:
M98 P"0:/macros/G29"
SD card not printing
ok.


I have a macro for pause and resume used for filament run out.  They work when filament out or from the printer lcd control.  But if I resume using my macro I get this error:
5/12/2020, 10:02:03 AM | M98 P"0:/macros/MY_RESUME"JSON.parse: unexpected non-whitespace character after JSON data at line 1 column 22 of the JSON data

**Comment by markwinger** [comment]
I also see a lot of messages: "response too long, see console"  but I see nothing in the console about them.

**Comment by markwinger** [comment]
Is there a more appropriate place to report issues I see?  There are many.

**Comment by pluuuk** [comment]
@markwinger, can you summarize all the issues you're seeing? I think this is the appropriate place to report issues. The ack system in gcode.py changed, so it's expected to see more messages than usual , but i haven't seen that many.

**Comment by pluuuk** [comment]
About the SDcard messages, it seems it's an issue with octoprint. Can you disable octoprint and test if the messages go away? 
https://github.com/OctoPrint/OctoPrint/issues/3258

**Comment by pluuuk** [comment]
I will take a look at the macro system and report back what i find.

**Comment by markwinger** [comment]
How do I disable octoprint?

**Comment by pluuuk** [comment]
It depends on your linux distro, but I'm guessing the commands down here should work. 

sudo systemctl stop octoprint
If you don't want it to start automatically:
sudo systemctl disable octoprint
You can re-enable it with:
sudo systemctl enable octoprint

**Comment by markwinger** [comment]
That definitely stopped the endless streams of notification.  

I know the macros dont work and I'll document here other things I find odd or not working.

**Comment by manu7irl** [comment]
You problably should start from scratch as I have no errors at all! I run 2
printers simultaneously. Everything is ok.

**Comment by pluuuk** [comment]
@markwinger, to help with the macro issue, can you attach your macro?

**Comment by markwinger** [comment]
Below is the my_pause macro.  The my_pause macro is called when out of filament and that works correctly.  I can call it from the lcd control panel but fails if I try to execute it from dwc2.  my_pause calls my_temps which I also included.

If I right click on the macro in the file management/Macro on any of the macros and click edit, I get an error:
0:/macros/MY_PAUSE

These macros are all in the printer.cfg file.  Do I to move them out of the printer.cfg file or something?

[gcode_macro my_pause]
gcode:
     ; relative XYZE
     my_temps
     G91
     M83

     ; retract filament of 25 mm up, move Z slightly upwards and
     G1 Z+5 E-10 F1000
     G1 E-50 F1000

     ; absolute XYZE
     M82
     G90

     ; move to a safe rest position, adjust as necessary
     G1 X0 Y0
     M117 Filament out
[gcode_macro my_temps]
variable_nozzle_temp: 0
gcode:
     # save nozzle temp
     SET_GCODE_VARIABLE MACRO=my_temps VARIABLE=nozzle_temp VALUE={printer.extruder.target}
     M104 S0 ; turn off hot end heater
     SET_IDLE_TIMEOUT timeout=20000

**Comment by markwinger** [comment]
About pause.  If I pause using the pause button it stops and resume restarts but doe no retraction or head positioning.  The docs for it say that pause executes the pause.g system macro.  Under system I see no macros, only config.g which is printer.cfg.   I'm guessing if I want to change the pause functionality I need to create a pause macro, the current behavior is a default and there is no macro.  Is that correct?
(I'm doing a 12hr print right now or I would just try it)

**Comment by pluuuk** [comment]
For the pause and resume behavior, the original repo had MOVE=0, which does
not move the head back to its position.

The GCODE to resume and move the head is:

RESTORE_GCODE_STATE NAME=DWC_BOTTON MOVE=1

 I think this is the default behavior for reprapfirmware. I will take a
look at your macro issue later today.

**Comment by tripstar76** [comment]
Try changing the macro name to capitals......MY_PAUSE.

**Comment by markwinger** [comment]
G29 is already caps and it does not work.  (long print still running so I cant change anything yet)

None of my macros can be executed or edited in the file management section.  Should I be able to edit them?

**Comment by tripstar76** [comment]
@pluuuk 

Is this a typo?
line 1074:

	def cmd_M32(self, gcmd):
		self.sdcard.cmd_M23(gcmd)

**Comment by Szeker** [comment]
I beleive, it is "UPLOAD AND START" = select file and start SD print, so first load the file => M23, than start printing => M24.

So NOT a typo.

**Comment by pluuuk** [comment]
@tripstar76 and @Szeker, can you provide your comments about the fixes on #75?
It's hard to keep track of the conversation.
@tripstar76, Szeker is correct. M24 starts the print, M32 is the reprapfirmware command that is natively implemented by klipper as M23+ M24. 
Regarding the macros, I can take a look and see if I broke something when re implementing the gcodes for macro management. I will provide feedback to @markwinger here.

**Comment by markwinger** [comment]
Changed the macro to MY_PAUSE.  No Change.

**Comment by pluuuk** [comment]
@markwinger , I have fixed the macro issue.
Please do a git pull of the latest commit on the fork in #75 .

**Comment by markwinger** [comment]
Again, I don't know how git works so what commands do I issue to pull from #75.

**Comment by pluuuk** [comment]
I'm assuming that you installed using my instructions, so simply running `git pull` works.

**Comment by markwinger** [comment]
Progress, I can now run macros. 
I still cannot edit from file management.  Is that supported?

**Comment by pluuuk** [comment]
I don't think that's ever been supported.

**Comment by markwinger** [comment]
Macros are working.  So I wanted to try get the pause/resume buttons to run custom pause/resume gcode.   Things you should know says:
For pause and resume macros you can use:
    kliper gcode macros pause_print resume_print cancel_print (not case sensitive)

So I simply changed my_pause, my_resume macro names to pause_print and resume_print, restarted firmware and stated a print.  While printing I paused the print.  The printer stopped and dwc2 pops up: connection interupted attempting to reconnect. The gui is continuously flashing with this error.  The  klipper log shows:
Exiting SD card print (position 3628)
Uncaught exception GET /rr_status?type=3 (192.168.1.118)
HTTPServerRequest(protocol='http', host='192.168.1.107:4750', method='GET', uri='/rr_status?type=3', version='HTTP/1.1', remote_ip='192.168.1.118')
Traceback (most recent call last):
  File "/home/pi/klippy-env/local/lib/python2.7/site-packages/tornado/web.py", line 1592, in _execute
    result = yield result
  File "/home/pi/klippy-env/local/lib/python2.7/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/home/pi/klippy-env/local/lib/python2.7/site-packages/tornado/concurrent.py", line 261, in result
    raise_exc_info(self._exc_info)
  File "/home/pi/klippy-env/local/lib/python2.7/site-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 272, in get
    self.web_dwc2.rr_status_3()
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 898, in rr_status_3
    "status": self.get_printer_status() ,
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 1359, in get_printer_status
    self.pause_macro()
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 1226, in pause_macro
    self.gcode_queue.append(parse_params('PAUSE_PRINT'))
NameError: global name 'parse_params' is not defined
DWC2:76510.02 - 192.168.1.118 - GET - /rr_connect?password=reprap&time=2020-5-13T8%3A28%3A31
Unhandled exception during run
Traceback (most recent call last):
  File "/home/pi/klipper/klippy/klippy.py", line 178, in run
    self.reactor.run()
  File "/home/pi/klipper/klippy/reactor.py", line 251, in run
    g_next.switch()
  File "/home/pi/klipper/klippy/reactor.py", line 278, in _dispatch_loop
    timeout = self._check_timers(eventtime)
  File "/home/pi/klipper/klippy/reactor.py", line 136, in _check_timers
    t.waketime = waketime = t.callback(eventtime)
  File "/home/pi/klipper/klippy/reactor.py", line 48, in invoke
    res = self.callback(eventtime)
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 1614, in update_printdata
    if self.get_printer_status() in ['S', 'P', 'D']:
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 1359, in get_printer_status
    self.pause_macro()
  File "/home/pi/klipper/klippy/extras/web_dwc2.py", line 1226, in pause_macro
    self.gcode_queue.append(parse_params('PAUSE_PRINT'))
NameError: global name 'parse_params' is not defined
~

**Comment by markwinger** [comment]
I tried a different approach.  I renamed pause_print, resume_print to my_pause_print and my_resume_print so they would not be executed but the pause/resume buttons.
I then created 2 dwc2 macros, pause.g and resume.g.  I tried clicking the pause button and it worked. So I clicked the resume button and it resumed immediately, but did not wait for nozzle temp to reach target, but it did start heating back up to the target temp.  Since I did not wait very long before resuming, the temp had not dropped much so extrusion temp was still above minimum.  So I repeated the test this time waiting for the temp to get well below minimum temp.
This time when I resumed, the buttons immediately changed to grayed pause and a start again button and showed print 100% done.  But the macro was running, because a the temperature target was reset, nozzle started to heat.  When target was reached I got a popup saying the idle timeout had been reset to 600(Last thing in the resume macro).

At this point printer showed print 100% complete with hotend and bed at target temperatures.  I waited 10 minutes to confirm idle timeout turned off heaters.  That worked.

So I repeated the process using the my_pause_printer and my_resume_print macros to see if they work correctly.  But got the error about white space in the json file again.

Rebooted, ran some macros ok, so tried again.  While printing, all macros report the white space error.

So the problem now appears to be macros work, but not while printing.

I attached my printer.cfg file.
[config.txt](https://github.com/Stephan3/dwc2-for-klipper/files/4622969/config.txt)

**Comment by markwinger** [comment]
I tried my filament runout renabling octoprint and it fails to start correctly that way as well.  How do I disable dwc2 to insure it is not interfering with octoprint?

**Comment by pluuuk** [comment]
You can rename the file in:

klipper/klippy/extras/dwc2-for-klipper.py to dwc2-for-klipper.py.bak

To enable it again, rename it back to the original extension.

**Comment by pluuuk** [comment]
> PAUSE_PRINT
@markwinger , I have pushed a fix to the issue in this comment, it was a typo in the pause gcode execution. It should work as nominal now.
Rename your macros to how you had them setup originally, don't list them as uppercase.

**Comment by pluuuk** [comment]
@markwinger , can you please run
```
cd ~/dwc2-for-klipper/
git pull
```

And retest?

**Comment by markwinger** [comment]
I have not tried your change yet.  But I have a new piece of info.  I did a pause, which saves the temp.  I did a resume which resets the temp with an M109.  The print resumed immediately but the temp target did not change. I waited a minute or two and did a pause again.  At this point the M109 must have been issued because the target changed back to the original temp.  When the target temp was reached, the resume took place.  This is really interesting.  Here is the resume macro.
 [gcode_macro RESUME_PRINT]
gcode:
    reset_temp
     SET_IDLE_TIMEOUT timeout=600
     resume

Notice the reset_temp is the first thing done.  Which should wait for target temperature before continuing.  But I see the SET_IDLE_TIMEOUT when the resume was done as if it skipped the reset_temp.  But then the reset_temp appears to run later when I run the pause.  Seems like we have an sequencing issues here.

I cant do the pull right now but hopefully I can in an hour or 2.

**Comment by markwinger** [comment]
I did your pull.  Although I don't know how your code works, it appears that the macros are forked or run in a different thread so the job will continue to run even though the macro is waiting.  This clear when I do a resume, the print starts immediately, not waiting for the macro to complete.  I arrive at this conclusion because when I do resume and pause again before the m109 of the previous macro is complete, the next macro does not start until the m109 is done.  i.e it keeps printing until the target temp is reached, then it pauses.  I also see the message saying the idle_timeout has changed reinforcing this conclusion. 

A different issue, if I pause the job 3 time is goes to done.  There must be some pause_resume limit imposed somewhere.

Also, I see popups occasionally saying something like "response too long.  See console"  But the console doesn't show it.

**Comment by pluuuk** [comment]
@markwinger , do the macros work when you run octoprint? I'm not familiar with the macro syntax, but why do you specify "reset_temp" and "resume" are those variables you have defined?

**Comment by markwinger** [comment]
Yes, the macros work in octoprint.
As the macros are current setup, I have a macro call PAUSE_PRINT, which is called by the filament runout code.  It is named this way so when I press the "Pause" in dwc2, it will be called.  The macro contains a call to another macro called my_temps.  That macro gets the current hotend target temp and stores in it in a macro variable called nozzle_temp, sets the hotend temp to 0, then sets the idle_timeout very long(so the hotbed does not get turned of an loose adhesion).  After return from my_temps, the PAUSE_PRINT does retraction, and moves the print head to 0,0.  Klipper then pauses.

Then I have RESUME_PRINT which is called by a menu selection.  Also because of the name it is called by dwc2 when the "Resume" button is clicked.  This macro calls macro reset_temps, which simply sends an M109 with the nozzle target temp saved by my_temps.  The M109 is supposed to pause until target temp is reached.  When reset_temps returns to RESUME_PRINT,  the idle_timeout is set back to 600 and resume is called telling klipper to resume.

The M109 not causing the print to wait, but another macro to pause is where my problem seems to be originating.

**Comment by markwinger** [comment]
One more issue I noticed.  After this testing I end up with the job done/cancelled/???  but the bed and nozzle heaters are still on.  So I tried going to tools->control-all-> everything off.  I get an error:

G10 P0 R-273.15 S-273.15 M140 P0 S-273.15
Error:  Requested temperature (-273.1) out of range (0.0:265.0)
Error:  Requested temperature (-273.1) out of range (0.0:130.0)

Wow, absolute 0.  That's cool.  Cooler that the climate here in Minnesota.
But if I set the temps to 0, it works correctly.

**Comment by changedsoul** [comment]
I must be missing something. I cannot get this to work for the life of me. There seems to be multiple commands that claim to work and nothing works for me. 
The only thing that gets me close is from here: https://klipper.info/klipper-+-dwc2-1/installing-dwc2
Yet I still get the "try_load_module" error. 
I found the suggestion to rename that in one of the previous posts, and this got me to see the web interface, yet I was still unable to connect. 
When I perform a "FIRMWARE_RESTART" I get the following error: Error:  'dict' object has no attribute 'get_commandline'

Any idea whats wrong?
This is all that is in /tmp/klipper.log after I issue the command:
```
"FIRMWARE_RESTART
DWC2:24782.43 - 10.0.0.44 - GET - /rr_gcode?gcode=FIRMWARE_RESTART
FIRMWARE_RESTART
failed: FIRMWARE_RESTART'dict' object has no attribute 'get_commandline'
"
```

**Comment by markwinger** [comment]
I think I know what my main problem is.  When you click the resume button, the printer is resumed then the macro is run.  In octoprint, the macro is run first.  So I tried creating resume.g macro, and renamed the macro RESUME_PRINT to MY_RESUME_PRINT so dwc2 does not run it.  I then created a dwc2 macro called my_resume_print, and ran it directly with out clicking the resume button.

The macro runs properly, but dwc2 continues to stay paused.  Temp goes up correctly, idle_timeout get set when temp hits target, but dwc2 does not resume.  Clipper however is in a different state.  The logs show:
DWC2:68106.65 - 192.168.1.118 - GET - /rr_config
passed: RESET_TEMP
idle_timeout: Timeout set to 600.00 s
passed: SET_IDLE_TIMEOUT
Print is not paused, resume aborted
passed: RESUME

So my macro worked, klipper is ready to go, but dwc2 is not resumed. 

So I have 2 problems, and they may not be solvable
1-dwc2 appears to run the macro after resuming print.  So I cannot use the resume button.
2-resume in a macro only resumes klipper and dwc2 does not seem to know we have resumed.  So unless we have a way to tell dwc2 to resume from a macro, this may just be impossible.
